### DISCUSSION TITLE
Запуск zapret в контейнере docker

### DISCUSSION THREAD

**Discussion by mpuchko** [bol-van/zapret#1008] [Q&A]
реально запустить zapret в контейнере docker под ubuntu 24?
1. нужен ли режим контейнера --network=host?
2. что отвечать на вопрос по iptables в скрипте установки? 

я смог запустить, но выдается ошибка, yb не работает:

контейнер создавал так:
```
docker run -it -p 1090:1090 ubuntu:latest /bin/bash
```

```
root@54f9312eee70:/tmp# ./blockcheck.sh
* checking system
firewall type unsupported not supported in Linux

press enter to continue
```

```
root@54f9312eee70:/tmp# /opt/zapret/init.d/sysv/zapret start
Starting daemon 2: /opt/zapret/tpws/tpws --user=tpws --socks --bind-addr=127.0.0.1 --bind-iface4=eth0 --bind-wait-ifup=30 --bind-wait-ip=30 --port=987
--filter-tcp=80 --methodeol  --new
--filter-tcp=443 --split-pos=1,midsld --disorder
Starting daemon 1: /opt/zapret/tpws/tpws --user=tpws --bind-addr=127.0.0.127 --port=988
--filter-tcp=80 --methodeol  --new
--filter-tcp=443 --split-pos=1,midsld --disorder
Starting daemon 3: /opt/zapret/nfq/nfqws --user=tpws --dpi-desync-fwmark=0x40000000 --qnum=200
--filter-tcp=80 --dpi-desync=fake,multisplit --dpi-desync-split-pos=method+2 --dpi-desync-fooling=md5sig  --new
--filter-tcp=443 --dpi-desync=fake,multidisorder --dpi-desync-split-pos=1,midsld --dpi-desync-fooling=badseq,md5sig  --new
--filter-udp=443 --dpi-desync=fake --dpi-desync-repeats=6
root@54f9312eee70:/tmp# setpcap: Operation not permitted
```

```
root@54f9312eee70:/tmp# ip -c a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
14: eth0@if15: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 172.17.0.2/16 brd 172.17.255.255 scope global eth0
       valid_lft forever preferred_lft forever
```

```
root@54f9312eee70:/tmp# ps faux
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
...
tpws        1952  0.0  0.0   1064   128 pts/1    Sl   22:06   0:00 /opt/zapret/tpws/tpws --user=tpws --socks --bind-addr=127.0.0.1 --bind-if
tpws        1972  0.0  0.0    400   128 pts/1    S    22:06   0:00 /opt/zapret/tpws/tpws --user=tpws --bind-addr=127.0.0.127 --port=988 --fi
```

**Comment by bol-van** [comment]
Докер непонятно зачем нужен. У zapret зависимостей практически нет
Судя по логу включено было все подряд на авось. И tpws, и nfqws. Так, для кучи, на всякий случай
Сочетание tpws и nfqws - продвинутая методика, без понимания как оно работает вряд ли стоит

**Reply by mpuchko** [reply]
докер нужен для изоляции в рамках контейнера. цель - создать socks5 proxy, перенаправляя через который трафик получаем эффект. а так все работает без изменений. глобально модифицировать сервер я опасаюсь, т.к. на нем важная информация и он нужен для работы.

что значит эта ошибка:
setpcap: Operation not permitted
?

**Reply by bol-van** [reply]
Наверно не хватает каких-то caps. Там кастрированный рут

**Reply by bol-van** [reply]
socks это одно дело, но другое дело tpws transparent и nfqws
и все что-то дурят
и сокс тоже дурит

каша получается

**Reply by bol-van** [reply]
не установлены ни nftables, ни iptables в контейнер

**Reply by spvkgn** [reply]
> что значит эта ошибка: setpcap: Operation not permitted ?

`--cap-add=NET_ADMIN --cap-add=NET_RAW`

**Reply by bol-van** [reply]
еще CAP_SETPCAP

**Comment by 8hrsk** [comment]
Приветствую! Можете ознакомится с моим реаозиторием [zapret-docker-proxy](https://github.com/8hrsk/zapret-docker-proxy)

В данный момент сам планирую переводить сервис на shadowsocks. Оптимизация и носки для нормальной работы Discord.

Проект достаточно простой. Думаю, разобраться проблемой не будет. В данный момент работает под Squid, но изменить не трудно. Буду рад, если создадите pull реквест с вашим решением <3

**Reply by 8hrsk** [reply]
А вообще, рекомендую использовать Dockerfile и docker-compose. В Dockerfile нужно прописать правильную настройку. Установить все нудные пакеты (iptables, curl, dnscrypt, ваш прокси), копировать конфиги для systemd.resolv и dnscrypt.

По сути, самое важное — правильно настроить среду. Одной коммандой это сделать не выйдет. Zapret очень комплексный и нуждается в хорошо отлаженной среде для работы

**Reply by bol-van** [reply]
В чем он комплексный ? Сидит в /opt/zapret.  От системы требует лишь curl и ip/nf tables.
Больше ему ничего не нужно

**Reply by 8hrsk** [reply]
Если разбираться и делать что-то свое, особенно, если с этими технологиями раньше дела не имел, то дело идет туго

**Reply by bol-van** [reply]
С докером пойдет еще туже. Потому что zapret все равно придется настраивать и еще придется изучать докер.  Придется прописывать socks в броузеры, эффект прозрачности утрачивается. Либо придется городить маршрутизацию через контейнер, что еще усложняет процесс в разы

**Reply by 8hrsk** [reply]
Ну с докером у меня проблем не было, ибо с ним я работаю часто. За два дня разобрался, как поднять zapret в докере без косяков. Сейчас только пытаюсь настроить всё так, чтобы нормально работал discord. Все остальные сервисы запускаются без проблем, а вот войсы в дискорде так и не хотят работать. Как я понимаю, это из-за того, что нужно заменить Squid на SOCKS прокси, чтобы нормально работали RPC соединения

**Reply by bol-van** [reply]
Войс трафик идет через udp, и его не настроить через прокси

**Reply by 8hrsk** [reply]
Тогла каким образом поступить? VPN сервер какой-нибудь поднять, вместо прокси? WireGuqrd, например

**Reply by bol-van** [reply]
Отказаться от докера - самое простое решение

**Reply by 8hrsk** [reply]
Вот тут и проблема. Я хочу поднять сервис, к которому я смогу подключиться как к прокси/впн откуда угодно с любого устройства. Докер в данном случае просто помогает изолировать среду от остальной развернутой на сервере инфраструктуры. Запрет в докере с прокси у меня работает прекрасно, скорость стабильная, блокировки обходит. Вот только голосовые каналы в дискорде нужно починить

**Reply by bol-van** [reply]
Тут есть статейка как это сделать без VPN. Но потребуется химичить на клиенте. Соответственно, мобильники отдохнут
https://habr.com/ru/articles/855164

Или правда делать VPN. Но VPN - это опять же маршрутизация, которую надо настраивать. Нужен VPN клиент на устройствах

**Reply by 8hrsk** [reply]
Ну впн клиент на устройстве сделать не так трудно. Есть много вариантов. Например, тот же Cisco. Айпишник у меня белый, доменное имя есть. Потом, как желание будет, заведу vps заграничный какой-ниудь, сделаю его впн сервером. А так, хочется сделать сейчас через запрет. Утилита удобная, работать с ней приятно.

В общем, поклон и уважение вам)

**Reply by bol-van** [reply]
Кто-то наоборот плюется и сыпет проклятиями. Вообщем, как обычно, на вкус и цвет

**Reply by 8hrsk** [reply]
Ну тут да, кому как. Но с другой стороны, никто никому ничего и не обязан. Проект открытый. Не нравится — переписывай)

А вообще хочу спросить, какой VPN лучше использовать? OpenVPN работает плохо, если вообще еще живой. WireGuard то ли блокируют, то ли нет. Есть еще Amnesia, про неё я ничего сказать не могу.

Думаю, брать в охапку WireGuard, но хочу спросить совета, если копались уже в этой теме

**Reply by bol-van** [reply]
amnezia - нормальная штука, если настроить параметры дурения протокола
Но придется собирать самому модуль ядра, что не очень удобно

**Reply by 8hrsk** [reply]
Звучит как-то не очень...

**Reply by bol-van** [reply]
Собирать только на linux.
На винде и телефонах не надо

**Comment by 8hrsk** [comment]
Режим host не нужен, пробросьте порт/порты сыоего проскси и все. Iptables - fw. Для начала нужно их поставить. Просто в dockerfile `apt install iptables`

**Comment by iKoalaa** [comment]
@bol-van @8hrsk Вопрос вроде по теме, так что напишу в этом топике.
Я запустил на маке контейнер вот такой контейнер https://github.com/vernette/ss-zapret/tree/master
Как прокси с стандартным маковским socks он, вроде как пускает через него трафик без особых проблем, но.
блокчек работает примерно никак, в том смысле что не может подобрать нормально стратегию.
Ковыряния и повтор стратегий, которые работают на винде, не работают в моей конфигурации(мак + докер + локальное прокси)

Буду рад если напишите мысли на тему, что ещё можно попробовать, или в каком месте может быть проблема.
